src/vs/workbench/contrib/notebook/browser/view/renderers/stdOutErrorPreProcessor.ts

```diff
@@ -0,0 +1,58 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { VSBuffer } from 'vs/base/common/buffer';
+import type { IOutputItemDto } from 'vs/workbench/contrib/notebook/common/notebookCommon';
+
+
+export function isStreamMime(mimeType: string) {
+	return ['application/vnd.code.notebook.stdout', 'application/x.notebook.stdout', 'application/x.notebook.stream', 'application/vnd.code.notebook.stderr', 'application/x.notebook.stderr'].includes(mimeType);
+}
+
+/**
+ * Given a stream of individual stdout outputs, this function will return the compressed lines, escaping some of the common terminal escape codes.
+ * E.g. some terminal escape codes would result in the previous line getting cleared, such if we had 3 lines and
+ * last line contained such a code, then the result string would be just the first two lines.
+ */
+export function compressOutputItemStreams(mimeType: string, outputs: IOutputItemDto[]) {
+	const buffers: Uint8Array[] = [];
+	let startAppending = false;
+
+	// Pick the first set of outputs with the same mime type.
+	for (const output of outputs) {
+		if (output.mime === mimeType) {
+			if ((buffers.length === 0 || startAppending)) {
+				buffers.push(output.data.buffer);
+				startAppending = true;
+			}
+		} else if (startAppending) {
+			startAppending = false;
+		}
+	}
+	compressStreamBuffer(buffers);
```

Comment:
We are doing `compressStreamBuffer` in serializer, so not sure if we need to do this here. The concern is `stream.subarray(0, MOVE_CURSOR_1_LINE_COMMAND.length).toString()` in a relatively large progress output might be slower than just copying the buffer and sending over to the webview.

Reply:
A manual test of 1000 lines of stdout output (1000 `appendOutputItem`) is probably sufficient

Reply:
>  to the webview (cost of toString() + GC of the short lived string).

Not following you here, there's no `toString` taking place here, we're sending the raw bytes